File: Grafo/uri1141.py
def cultivando(lista, peso, i, num):
    if(peso[i] != 0):
        return peso[i]
    for j in range(i + 1, num):
        max = peso[i]
        if(lista[j].find(lista[i]) != -1):
            #print(lista[j] + " : " + lista[i])
            temp = 1 + cultivando(lista, peso, j, num)
            if(temp > max):
                max = temp
            peso[i] = max
        #print("for : i: " + str(i) + " j: " + str(j))
        #print(str(peso[i]) + " " + str(peso[j]))


    #max = 0
    #for j in range(i + 1, num):
    #    if(peso[j] > max and lista[j].find(lista[i]) != -1):
    #        max = peso[j]
    #print("rec : i: " + str(i))
    #print(str(peso[i]))
    return peso[i]

while(True):
    num = int(input())
    if(num == 0):
        break
    list = []
    peso = []
    for i in range(num):
        string = input()
        list.append(string)
        peso.append(0)
    list = sorted(list, key = lambda x: len(x))

    max = 0
    for i in range(num):
        temp = cultivando(list, peso, i, num)
        if(temp > max):
            max = temp
    #for i in range(len(peso)):
    #    print(peso[i])
    #    print(list[i])
    print(max + 1)

File: Grafo/uri1141.cpp
#include <bits/stdc++.h>

using namespace std;

int num;
char **vet;
int *peso;

bool compare(char *str1, char *str2){
	return strlen(str1) < strlen(str2);
}

int cultivando(int i, int num){
	if(peso[i] != 0){
		return peso[i];
	}
	for(int j = i + 1; j < num; j++){
		int max = peso[i];
		if(strstr(vet[j], vet[i]) != NULL){
			int temp = 1 + cultivando(j, num);
			if(temp > max){
				max = temp;
			}
			peso[i] = max;
		}
	}
	return peso[i];
}

int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	while(1){
		num;
		scanf("%d", &num);
		if(num == 0) break;
		vet = (char **) malloc(sizeof(char *) * num);
		peso = new int[num];
		for(int i = 0; i < num; i++){
			getchar();
			char str[1001];
			scanf("%s", str);
			vet[i] = (char *) malloc(sizeof(char) * (strlen(str) + 1));
			strcpy(vet[i], str);
			peso[i] = 0;
		}
		sort(vet, vet + num, compare);
		int max = 0;
		for(int i = 0; i < num; i++){
			int temp = cultivando(i, num);
			if(temp > max){
				max = temp;
			}
		}
		printf("%d\n", max + 1);
	}
}